src/libponyc/codegen/genfun.c

```diff
@@ -97,8 +97,14 @@ static void make_signature(compile_t* c, reach_type_t* t,
   size_t tparam_size = count * sizeof(LLVMTypeRef);
 
   if(message_type)
+  {
     tparam_size += tparam_size + (2 * sizeof(LLVMTypeRef));
 
+    // must be at least the size of 3 or else we get a heap buffer overflow
+    if((2 * sizeof(LLVMTypeRef)) == tparam_size)
+      tparam_size += tparam_size + sizeof(LLVMTypeRef);
+  }
```

Comment:
what's going on here. Joe and I are trying to sort it out and we are very confused at the moment.

Reply:
i assume this was found by the sanitizer? 

either way, i find the logic confusing. is there a reason it feels so "obfuscated"?

Reply:
going to ignore you @SeanTAllen because @jemc rewrote the code already..